### DISCUSSION TITLE
How to change baseFeePerGas in testing

### DISCUSSION THREAD

**Discussion by RippleLeaf** [ConsenSys-archive/truffle#5265] [General Q&A] [answered]
I am testing a smart contract on a local development chain. The smart contract runs normally with ganache CLI 6.12.2 (ganache-core: 2.13.2). After updating ganache to the latest version (7.3.2), with `truffle test` it gives the following error:

`Error: while migrating Migrations: Returned error: VM Exception while processing transaction: Transaction's maxFeePerGas (0) is less than the block's baseFeePerGas (875000000) (vm hf=london -> block -> tx)`

May I know if there is a way to set `baseFeePerGas=0` to zero for testing? I have tried setting it in `truffle-config.js` and in the command line but neither worked, and I didn't find any online document about how to set it.

**Comment by RippleLeaf** [comment]
@MicaiahReid Hi, I opened this discussion question here. Thank you for your reply in this [issue](https://github.com/ethereum/go-ethereum/issues/24661)!

**Reply by MicaiahReid** [reply]
Thanks for opening this here! I'm sure others will have the same question, so it's nice to have it formally answered where others can find it!

**Comment by MicaiahReid** [comment, marked as answer]
@RippleLeaf We don't currently have a way to set the `baseFeePerGas`, but we do have [an issue for it](https://github.com/trufflesuite/ganache/issues/1573), so you can expect to see this feature in the near(ish) future.

In terms of workarounds, you have a few options:
1. Update your tests to set appropriate an `gasPrice` on your transactions. However, I obviously don't know the details of what you're working on, so this may not be feasible or helpful.
2. Start Ganache with the `--chain.hardfork berlin` option. This will use berlin hardfork rules for the chain, which was before blocks had a `baseFeePerGas`.
3. Use `evm_mine` to lower the `baseFeePerGas` of blocks. Every time a block is mined, the `baseFeePerGas` is either lowered or raised based off of how full the previous block is. The max that the `baseFeePerGas` can be raised or lowered from one block to the next is 12%. So if you used `evm_mine` to mine 125 blocks on startup, the `baseFeePerGas` would lower from 875000000 WEI to pretty close to 0 WEI.

**Reply by RippleLeaf** [reply]
Thank you very much! I run with the berlin hardfork and it works fine now. 

The new version of ganache somehow also solved another problem I had earlier - I guess it was something related to memory overflow, but I didn't find the exact reason. Now the line causing transaction revert in the earlier version of ganache just works fine... I still don't know why but it is good! Thank you for your help again!

**Reply by staminna** [reply]
@MicaiahReid Can you elaborate the commands/files need to be adjusted to set the evm_mine and baseFeePerGas? I am not sure how to calculate values ideal for the app. I was able to do two transactions locally until it exhausted the gas on my Ganache (localhost).

**Reply by MicaiahReid** [reply]
Hi @staminna! To clarify, this method is only really helpful if you're trying to _lower_ a block's `baseFeePerGas`, and not to an exact value. 
Something like this would repeatedly mine an empty block until the `baseFeePerGas` is lowered below a desired value:
```TypeScript
  const provider = ganache.provider(options);
  const DESIRED_BASE_FEE =  100000000; // set to whatever you want
  let block;
  do {
    block = await provider.request({
      method: "eth_getBlockByNumber",
      params: ["latest", true]
    });
    console.log(block.baseFeePerGas);
    await provider.request({
      method: "evm_mine",
      params: []
    });
  } while (parseInt(block.baseFeePerGas) > DESIRED_BASE_FEE);
```
